api-put-object-streaming.go

```diff
@@ -343,13 +346,30 @@ func (c Client) putObjectNoChecksum(ctx context.Context, bucketName, objectName
 		}
 	}
 
+	var md5Base64 string
+	if opts.SendContentMd5 {
+		// Create a buffer.
+		buf := make([]byte, size)
+		defer debug.FreeOSMemory()
```

Comment:
`runtime/debug` and `defer debug.FreeOSMemory()` is already a part of the SDK ([](https://github.com/minio/minio-go/blob/master/api-put-object-multipart.go#L101)) , it helps ensure memory is freed up. Not sure why to remove this

Reply:
Didn't notice that, my bad

Reply:
`buf := make([]byte, size)`  this line causes a panic on large files ( tried a 7GB file here)

```
fatal error: runtime: out of memory

runtime stack:
runtime.throw(0x7680d4, 0x16)
        /usr/local/go/src/runtime/panic.go:774 +0x72
runtime.sysMap(0xc004000000, 0x1c0000000, 0xa0c1d8)
        /usr/local/go/src/runtime/mem_linux.go:169 +0xc5
runtime.(*mheap).sysAlloc(0x9f3c00, 0x1c0000000, 0x66d9ca, 0x17aeb1)
        /usr/local/go/src/runtime/malloc.go:701 +0x1cd
runtime.(*mheap).grow(0x9f3c00, 0xe0000, 0xffffffff)
        /usr/local/go/src/runtime/mheap.go:1252 +0x42
runtime.(*mheap).allocSpanLocked(0x9f3c00, 0xe0000, 0xa0c1e8, 0x1f)
        /usr/local/go/src/runtime/mheap.go:1163 +0x291
runtime.(*mheap).alloc_m(0x9f3c00, 0xe0000, 0x7ff2e3950101, 0x7ff2e395ddd8)
        /usr/local/go/src/runtime/mheap.go:1015 +0xc2
runtime.(*mheap).alloc.func1()
        /usr/local/go/src/runtime/mheap.go:1086 +0x4c
runtime.(*mheap).alloc(0x9f3c00, 0xe0000, 0x10101, 0x7ff2e3909028)
        /usr/local/go/src/runtime/mheap.go:1085 +0x8a
runtime.largeAlloc(0x1c0000000, 0xa00101, 0x7ff2e3909028)
        /usr/local/go/src/runtime/malloc.go:1138 +0x97
runtime.mallocgc.func1()
        /usr/local/go/src/runtime/malloc.go:1033 +0x46
runtime.systemstack(0x458d24)
        /usr/local/go/src/runtime/asm_amd64.s:370 +0x66
runtime.mstart()
        /usr/local/go/src/runtime/proc.go:1146

goroutine 1 [running]:
runtime.systemstack_switch()
        /usr/local/go/src/runtime/asm_amd64.s:330 fp=0xc00012aed0 sp=0xc00012aec8 pc=0x458e20
runtime.mallocgc(0x1c0000000, 0x6ecc80, 0x764001, 0xc000085680)
        /usr/local/go/src/runtime/malloc.go:1032 +0x895 fp=0xc00012af70 sp=0xc00012aed0 pc=0x40d935
runtime.makeslice(0x6ecc80, 0x1c0000000, 0x1c0000000, 0x7ff2e3909028)
        /usr/local/go/src/runtime/slice.go:49 +0x6c fp=0xc00012afa0 sp=0xc00012af70 pc=0x443d5c
github.com/minio/minio-go/v6.Client.putObject(0xc0000d6400, 0xc0000647e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0000855f0, 0xc00000e840, ...)
        /home/kp/code/src/github.com/minio/minio-go/api-put-object-streaming.go:352 +0x432 fp=0xc00012b598 sp=0xc00012afa0 pc=0x6a94a2
github.com/minio/minio-go/v6.Client.putObjectCommon(0xc0000d6400, 0xc0000647e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0000855f0, 0xc00000e840, ...)
        /home/kp/code/src/github.com/minio/minio-go/api-put-object.go:204 +0x46c fp=0xc00012b878 sp=0xc00012b598 pc=0x6abcdc
github.com/minio/minio-go/v6.Client.PutObjectWithContext(0xc0000d6400, 0xc0000647e0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0000855f0, 0xc00000e840, ...)
        /home/kp/code/src/github.com/minio/minio-go/api-put-object-context.go:32 +0x198 fp=0xc00012ba58 sp=0xc00012b878 pc=0x6a3708
github.com/minio/minio-go/v6.Client.PutObject(...)
        /home/kp/code/src/github.com/minio/minio-go/api-put-object.go:166
main.main()
        /home/kp/code/src/github.com/minio/minio-go/examples/s3/putobject.go:54 +0x42c fp=0xc00012bf60 sp=0xc00012ba58 pc=0x6ba00c
runtime.main()
        /usr/local/go/src/runtime/proc.go:203 +0x21e fp=0xc00012bfe0 sp=0xc00012bf60 pc=0x43055e
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1357 +0x1 fp=0xc00012bfe8 sp=0xc00012bfe0 pc=0x45aef1
exit status 2
```

Reply:
I am not sure there is something we can do here. To avoid this situation users should use the multi part upload (provide a relevant partsize as a part of PutObject call) - so minio-go only uses buffers equal to the size of partSize

Reply:
If SendContentMD5 is set can we add a validation error if it qualifies for multipart upload and partsize is not specified rather than failing with a panic?

Reply:
@poornas I changed the behaviour to allow multipart uploads when size is known as well. This means the for uploads greater than `partSize`, minio-go will try to use multipart uploads. This should fix the initial panic issue you reported.